### PR TITLE
Add `DfTx` PoolUpdatePair

### LINKED ISSUE #108: DeFi CustomTx scripting
> This tracks all the DeFi scripting (AKA CustomTx) that is to be implemented in jellyfish-transaction.

```
packages/jellyfish-transaction/src/script/defi/*
```

All CustomTX has a starting signature of `DfTx` in utf8 or `0x44665478`. It is as it is, don't need to convert from BE to LE.

They are structured as a 
1. `VarUInt` + [`OP_RETURN` + `DfTx` + `TYPES[CxTxType]` + ...]
1. `VarUInt` + `6a` + `44665478` + Type + ...

- [x] createMasternode: `'C'` @ivan-zynesis #283
- [x] resignMasternode: `'R'` @ivan-zynesis #283
- [x] createToken: `'T'` @canonbrother #208
- [x] mintToken: `'M'` @canonbrother #170
- [x] updateToken: `'N'` @canonbrother #300
- [x] updateTokenAny: `'n'` @canonbrother #300
- [x] createPoolPair: `'p'` @canonbrother #233
- [x] updatePoolPair: `'u'` @canonbrother #314 
- [x] poolSwap: `'s'` @fuxingloh #143
- [x] addPoolLiquidity: `'l'` @fuxingloh #145
- [x] removePoolLiquidity: `'r'` @fuxingloh  #146
- [x] utxosToAccount: `'U'` @ivan-zynesis #175 
- [x] accountToUtxos: `'b'` @ivan-zynesis #175 
- [x] accountToAccount: `'B'` @ivan-zynesis #175 
- [x] setGovVariable: `'G'` @ivan-zynesis #298
- [x] anyAccountsToAccounts: `'a'` @ivan-zynesis #181
- [x] AutoAuthPrep: `'A'`, @ivan-zynesis #279

### PR DESCRIPTION
**Pull request by canonbrother**
#### What kind of PR is this?:

/kind feature

#### Which issue(s) does this PR fixes?:

Fixes part of #108

### CODE REVIEW

File: packages/jellyfish-transaction/src/buffer/buffer_composer.ts [hunk 1]
```diff
@@ -498,4 +498,23 @@ export abstract class ComposableBuffer<T> implements BufferComposer {
       }
     }
   }
+
+  /**
+   * Unsigned Boolean 4 bytes, 1 = true, 0 = false
+   *
+   * @param getter to read from to buffer
+   * @param setter to set to from buffer
+   * @returns
```
Comment (fuxingloh) [inline packages/jellyfish-transaction/src/buffer/buffer_composer.ts:507]:
```suggestion
```

File: packages/jellyfish-transaction/src/buffer/buffer_composer.ts [hunk 2]
```diff
@@ -498,4 +498,23 @@ export abstract class ComposableBuffer<T> implements BufferComposer {
...
+    return {
+      fromBuffer: (buffer: SmartBuffer): void => {
+        setter(buffer.readInt32BE() === 1)
+      },
+      toBuffer: (buffer: SmartBuffer): void => {
+        var v = getter().toString().toLowerCase() === 'true' ? 1 : 0
+        buffer.writeBuffer(Buffer.from([0, 0, 0, v]))
+      }
+    }
+  }
```
Comment (fuxingloh) [inline packages/jellyfish-transaction/src/buffer/buffer_composer.ts:519]: Should isolated write tests for this

**Review by fuxingloh — CHANGES_REQUESTED** [review]

File: packages/jellyfish-transaction/src/buffer/buffer_composer.ts [hunk 3]
```diff
@@ -498,4 +498,22 @@ export abstract class ComposableBuffer<T> implements BufferComposer {
...
+   * @param getter to read from to buffer
+   * @param setter to set to from buffer
+   */
+  static uBool32 (getter: () => boolean, setter: (data: boolean) => void): BufferComposer {
+    return {
+      fromBuffer: (buffer: SmartBuffer): void => {
+        setter(buffer.readInt32BE() === 1)
+      },
+      toBuffer: (buffer: SmartBuffer): void => {
+        var v = getter().toString().toLowerCase() === 'true' ? 1 : 0
```
Comment (fuxingloh) [inline packages/jellyfish-transaction/src/buffer/buffer_composer.ts:514]:
```suggestion
        const v = getter() === true ? 1 : 0
```

Maybe should not do type coercion here?

File: packages/jellyfish-transaction/src/script/defi/dftx_pool.ts [hunk 4]
```diff
@@ -104,3 +105,51 @@ export class CPoolRemoveLiquidity extends ComposableBuffer<PoolRemoveLiquidity>
...
+              p.customRewards.push(new CTokenBalance(buffer).toObject())
+            }
+          }
+        },
+        toBuffer: (buffer: SmartBuffer): void => {
+          if (p.customRewards !== undefined) {
+            writeVarUInt(p.customRewards.length, buffer)
+            p.customRewards.forEach(data => new CTokenBalance(data).toBuffer(buffer))
+          }
+        }
```
Comment (fuxingloh) [inline packages/jellyfish-transaction/src/script/defi/dftx_pool.ts:151]: kinda sad that we need to do this, but I guess we don't really have a choice.

**Review by fuxingloh — COMMENTED** [review]
1 comment about type coercion, maybe should not do it?

File: packages/jellyfish-transaction/src/script/defi/dftx_pool.ts (hunk 4 above)
Comment (canonbrother) [reply packages/jellyfish-transaction/src/script/defi/dftx_pool.ts:151]: yea this is the only way i can think of for now
and i applied the same on `createpoolpair` as well
https://github.com/DeFiCh/jellyfish/pull/233

**Review by fuxingloh — APPROVED** [review]

File: packages/jellyfish-transaction/src/buffer/buffer_composer.ts [hunk 5]
```diff
@@ -498,4 +498,22 @@ export abstract class ComposableBuffer<T> implements BufferComposer {
...
+   * @param getter to read from to buffer
+   * @param setter to set to from buffer
+   */
+  static uBool32 (getter: () => boolean, setter: (data: boolean) => void): BufferComposer {
+    return {
+      fromBuffer: (buffer: SmartBuffer): void => {
+        setter(buffer.readInt32BE() === 1)
+      },
+      toBuffer: (buffer: SmartBuffer): void => {
+        const v = getter() ? 1 : 0
```
[the remark this answers is not in the document]
Comment (canonbrother) [reply packages/jellyfish-transaction/src/buffer/buffer_composer.ts:514]: yes you're right

[the remark this answers is not in the document]
Comment (canonbrother) [reply packages/jellyfish-transaction/src/buffer/buffer_composer.ts:514]: ohh.. just found out this `const v = getter() ? 1 : 0` is refined by `npm run standard`

**Review by monstrobishi — APPROVED** [review]

File: packages/jellyfish-transaction/src/buffer/buffer_composer.ts (hunk 5 above)
[the remark this answers is not in the document]
Comment (ivan-zynesis) [reply packages/jellyfish-transaction/src/buffer/buffer_composer.ts:514]: because TS (not just JS) do treat null and undefined as `false`

**Pull request merged**